Add canonical compiler examples

mhevery opened the pull request:
This PR is a beginning of canonical compiler examples (What do we expect the compiler to generate).

While doing this PR I ran into some issue and was forced to refactor:

1. Directives need to be passed into the `E` instructions since instantiation order can't be determined without breaking locality principle. Current implementation is a bit of a hack, follow on PR will create proper implementation.

2. The way queries register the local names is not consistent with the locality rule. In process of making new API, currently all of the query tests are broken. The local name `#local-name` on the element and the `exportAs` have to be resolved at runtime. /cc @pkozlowski-opensource

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -990,7 +1019,11 @@ export function lifecycle(lifeCycle: LifecycleHook, self?: any, method?: Functio
+    template?: ComponentTemplate<any>, 
+    tagName?: string, 
+    attrs?: string[],
     localName?: string): void {
```
Shouldn't we unify this with `elementStart` (that is, if `localName` goes away from `elementStart` and is moved to attributes, it should go away from here as well, right?)

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -396,29 +396,35 @@ export function injectViewContainerRef(): ViewContainerRef {
  */
 export function elementStart(
-    index: number, nameOrComponentDef?: string | ComponentDef<any>, attrs?: string[] | null,
-    localName?: string): RElement {
```
Is your intention to merge `localName` into `attrs`? If so we will have to check for the `#` prefix at runtime, right?

Alternative approach would be to have `localNames: string[]` and have all the local names (element, component and directives) in there. This way we have clear distinction between attributes and local names.

mhevery replied:
going with `string[]`

pkozlowski-opensource commented on packages/core/src/render3/assert.ts:
```diff
@@ -21,7 +21,10 @@
  */
 function stringifyValueForError(value: any): string {
-  return typeof value === 'string' ? `"${value}"` : '' + value;
+  if (value && value.native && value.native.outerHTML) {
```
Maybe create a dedicated function for stringifying `LNode`s instead?

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -466,9 +494,15 @@ function setUpAttributes(native: RElement, attrs: string[]): void {
-        native.setAttribute(attrs[i], attrs[i | 1]);
+    const key = attrs[i];
+    const value = attrs[i|1];
+    if (key.charAt(0) == '#') {
```
Is your intention to have any code here or just skip attributes representing local names? If it is to skip, just flip the condition. Otherwise add info / comment / implementation here.

mhevery replied:
reverted

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -396,29 +396,35 @@ export function injectViewContainerRef(): ViewContainerRef {
  * @param attrs Statically bound set of attributes to be written into the DOM element on creation.
- * @param localName A name under which a given element is exported.
  *
  * Attributes are passed as an array of strings where elements with an even index hold an attribute
```
If your intention is to merge attributes and local names arrays, please update this part of the doc accordingly (but also please check my other comments on merging 2 collections)

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -831,56 +865,51 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
+}
+
+export function directiveCreate<T>(
+    index: number, directive: T, directiveDef: DirectiveDef<T>, localName?: string): T {
```
I guess the `localName` is not the things here any more (logic to be moved to the `elementStart`?)

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -446,6 +455,25 @@ export function elementStart(
+        // is not guaranteed. Must be refactored to take it into account.
+        directiveCreate(++index, hostComponentDef.n(), hostComponentDef);
+      }
+      if (directiveTypes) {
```
If we embed this logic in `E(...)` all of the sudden it needs to know about all the directives processing logic (instantiate, set inputs, do things in the order required by DI, ...) and local names logic. This will make a simple hello world app pulling more code than before, if I'm not mistaken. Did you check the impact of this change on the "Hello, World" app size?

An alternative would be to have another instruction (sth like `InstantiateDirectives([Directive1Type, Directive2Type, ...])`. I know that this would have minimal impact on real-life apps (as you need directives vast majority of the time) but still, would be nice to keep instructions small and independent (as much as possible...)

mhevery committed:
refactor(ivy): move directive into elementStart

We used to have a separate `directive` instruction for instantiating
directives. However, such an instruction requires that directives
are created in the correct order, which would require that template
compiler would have knowledge of all dependent directives. This
would break template compilation locality principle.

This change only changes the APIs to expected form but does
not change the semantics. The semantics will need to be corrected
in subsequent commits. The semantic change needed is to
resolve the directive instantiation error at runtime based on
injection dependencies.

mhevery committed:
test(ivy): add canonical template translation examples

This change creates a spec file which contains canonical examples
of how the template compiler will translate templates into expected
output.

mhevery commented:
The code currently contains `hack_declareDirectives` and `hack_findQueryName` functions which can't be correctly implemented until @kara fixes refactors the data structures and gives us a place to store these values.

pkozlowski-opensource commented on packages/core/test/render3/compiler_canonical_spec.ts:
```diff
@@ -0,0 +1,135 @@
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['class', 'my-app', 'title', 'Hello'];
+      const MyComponentDef = defineComponent({
```
Why the `MyComponentDef` isn't a static prop of `MyComponent` - is it intentional?

mhevery replied:
changed

pkozlowski-opensource commented on packages/core/test/render3/compiler_canonical_spec.ts:
```diff
@@ -0,0 +1,135 @@
+      class MyComponent {
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['class', 'my-app', 'title', 'Hello'];
```
It will also need some file-wide index attached to the name as we can have multiple components in one file.

mhevery replied:
existing compiler has mechanism to deal with it. This is meant so we know what instructions to generate mainly.

pkozlowski-opensource commented on packages/core/test/render3/compiler_canonical_spec.ts:
```diff
@@ -0,0 +1,135 @@
+      }
+      // Important: keep arrays outside of function to not create new instances.
+      const e0_attrs = ['some-directive', ''];
+      const e0_dirs = [SomeDirective];
```
Both `const` need some file-wide index in the name

mhevery replied:
👍

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -427,30 +435,83 @@ export function elementStart(
+ * This function returns the queryName for a directive. It is a hack since we should
+ * compute the query value only once and store it with the template (rather than on each invocation)
+ */
+function hack_findQueryName(
```
I know this is a "hack" and will go away, but could we rename it to something that doesn't have `query` in the name? Essentially what we are doing here is to connect local refs with directive instances so maybe `hack_mapLocalnameToDirectiveInstance` ?

mhevery replied:
But that name is only used for `query` and nothing else. So I think the current name is right? Who else would use this value?

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -427,30 +435,83 @@ export function elementStart(
       node = createLNode(index, LNodeFlags.Element, native, componentView);
 
+      // TODO(misko): implement code which caches the local reference resolution
+      const queryName: string|null = hack_findQueryName(hostComponentDef, localRefs, '');
```
`queryName` => `localName` or `localNameForComponent`? Let's keep it independent from queries.

mhevery replied:
see previous comment. I think this only makes sense for queries. Who else would use this information.

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -830,57 +905,47 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
  *        be created or retrieved out of order.
  * @param directive The directive instance.
  * @param directiveDef DirectiveDef object which contains information about the template.
+ * @param queryName Name under which the query can retrieve the directive instance.
```
I would rename it to sht query independent, ex: `localName` or `nameInATemplate` - a name under which a directive instance is available in a template. Today query is making use of it name but it also declares names in a template and in the future we might have other usages (ex. animation queries).

tl;dr; my proposal is to keep this name query-independent.

mhevery replied:
see above

pkozlowski-opensource commented on packages/core/src/render3/instructions.ts:
```diff
@@ -830,57 +905,47 @@ export function textBinding<T>(index: number, value: T | NO_CHANGE): void {
+      ngDevMode &&
+          assertNotNull(previousOrParentNode.staticData, 'previousOrParentNode.staticData');
+      const nodeStaticData = previousOrParentNode !.staticData !;
+      (nodeStaticData.localNames || (nodeStaticData.localNames = [])).push(queryName, index);
```
I need to think a bit more about it, but I've got an impression that now we could skip saving  `localNames` is static data altogether. Previously we had to save a mapping between template-local names and directive instances since we haven't had this info on an element level (it was spread between element and directive instructions). But now everything is in one place.

So, maybe we can move logic of adding elements to queries in `elementStart` / `elementEnd`. This would let us store less info in memory.

mhevery replied:
follow up PRs welcome.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -427,30 +435,83 @@ export function elementStart(
+    // is not guaranteed. Must be refactored to take it into account.
+    for (let i = 0; i < directiveTypes.length; i++) {
+      // MEGAMORPHIC: `ngDirectiveDef` is a megamorphic property access here.
+      // This is OK, since we will refactor this code and store the result in `TViewState.data`
```
fixed

mhevery committed:
refactor(ivy): merged containerStart/containerEnd

This separation is no longer needed since directives are now passed into the `container` as an array rather than as child functions of the `containerStart`

mhevery committed:
refactor(ivy): remove `type` from `DirectiveDef`

This change makes the code cleaner for the user. It does mean
a little bit more work for us since we have to patch the `type` back
into the `DirectiveDef`. However since the patching happens only once
on startup it should not be significant.

kara reviewed (approved):
LGTM
